File: src/api/java/Probe.java
/**
Copyright (c) 2012-2014 Microsoft Corporation
   
Module Name:

    Probe.java

Abstract:

Author:

    @author Christoph Wintersteiger (cwinter) 2012-03-15

Notes:
    
**/ 

package com.microsoft.z3;

/**
 * Probes are used to inspect a goal (aka problem) and collect information that
 * may be used to decide which solver and/or preprocessing step will be used.
 * The complete list of probes may be obtained using the procedures
 * {@code Context.NumProbes} and {@code Context.ProbeNames}. It may
 * also be obtained using the command {@code (help-tactic)} in the SMT 2.0
 * front-end.
 **/
public class Probe extends Z3Object {
    /**
     * Execute the probe over the goal.
     * 
     * @return A probe always produce a double value. "Boolean" probes return
     *         0.0 for false, and a value different from 0.0 for true.
     * @throws Z3Exception 
     **/
    public double apply(Goal g)
    {
        getContext().checkContextMatch(g);
        return Native.probeApply(getContext().nCtx(), getNativeObject(),
                g.getNativeObject());
    }

    Probe(Context ctx, long obj)
    {
        super(ctx, obj);
    }

    Probe(Context ctx, String name) {
        super(ctx, Native.mkProbe(ctx.nCtx(), name));
    }

    @Override
    void incRef() {
        Native.probeIncRef(getContext().nCtx(), getNativeObject());
    }

    @Override
    void addToReferenceQueue() {
        getContext().getProbeDRQ().storeReference(getContext(), this);
    }
}

File: src/api/dotnet/Probe.cs
/*++
Copyright (c) 2012 Microsoft Corporation

Module Name:

    Probe.cs

Abstract:

    Z3 Managed API: Probes

Author:

    Christoph Wintersteiger (cwinter) 2012-03-21

Notes:
    
--*/

using System;
using System.Runtime.InteropServices;
using System.Diagnostics.Contracts;

namespace Microsoft.Z3
{
    /// <summary>  
    /// Probes are used to inspect a goal (aka problem) and collect information that may be used to decide
    /// which solver and/or preprocessing step will be used.
    /// The complete list of probes may be obtained using the procedures <c>Context.NumProbes</c>
    /// and <c>Context.ProbeNames</c>.
    /// It may also be obtained using the command <c>(help-tactic)</c> in the SMT 2.0 front-end.
    /// </summary>
    [ContractVerification(true)]
    public class Probe : Z3Object
    {
        /// <summary>
        /// Execute the probe over the goal. 
        /// </summary>
        /// <returns>A probe always produce a double value.
        /// "Boolean" probes return 0.0 for false, and a value different from 0.0 for true.</returns>
        public double Apply(Goal g)
        {
            Contract.Requires(g != null);

            Context.CheckContextMatch(g);
            return Native.Z3_probe_apply(Context.nCtx, NativeObject, g.NativeObject);
        }

        /// <summary>
        /// Apply the probe to a goal.
        /// </summary>
        public double this[Goal g]
        {
            get
            {
                Contract.Requires(g != null);

                return Apply(g);
            }
        }

        #region Internal
        internal Probe(Context ctx, IntPtr obj)
            : base(ctx, obj)
        {
            Contract.Requires(ctx != null);
        }
        internal Probe(Context ctx, string name)
            : base(ctx, Native.Z3_mk_probe(ctx.nCtx, name))
        {
            Contract.Requires(ctx != null);
        }

        internal class DecRefQueue : IDecRefQueue
        {
            public DecRefQueue() : base() { }
            public DecRefQueue(uint move_limit) : base(move_limit) { }
            internal override void IncRef(Context ctx, IntPtr obj)
            {
                Native.Z3_probe_inc_ref(ctx.nCtx, obj);
            }

            internal override void DecRef(Context ctx, IntPtr obj)
            {
                Native.Z3_probe_dec_ref(ctx.nCtx, obj);
            }
        };

        internal override void IncRef(IntPtr o)
        {
            Context.Probe_DRQ.IncAndClear(Context, o);
            base.IncRef(o);
        }

        internal override void DecRef(IntPtr o)
        {
            Context.Probe_DRQ.Add(o);
            base.DecRef(o);
        }
        #endregion
    }
}
